x/nexus/keeper/msg_dispatcher.go

```diff
@@ -63,9 +63,14 @@ func (m Messenger) routeMsg(ctx sdk.Context, msg exported.WasmMessage) error {
 		return err
 	}
 
+	// Prevent wasm gateway from spoofing a chain name registered in nexus as the source chain
+	if _, ok := m.GetChain(ctx, msg.SourceChain); ok {
+		return fmt.Errorf("wasm source chain %s should not be a registered chain in nexus", msg.SourceChain)
+	}
```

Comment:
While this is a good sanity check, I am not sure this is the best place for the check. It would be better to somehow handle this at chain registration. Also, out of curiosity, what would be the harm in allowing the same chain name to be registered in both places?

Reply:
> It would be better to somehow handle this at chain registration.
Checking at chain registration doesn't prevent the wasm layer to spoof it dynamically when routing a GMP call to core. This ensures that it's never possible.

> Also, out of curiosity, what would be the harm in allowing the same chain name to be registered in both places?
The chain name is part of the root of trust for a GMP call. Apps whitelisting by chain (e.g. ITS) are relying on a certain standard of security for the chain connection. The connection type can be completely different between core and amplifier for the same chain, so they should be distinguished in GMP by different identifiers. And in the event of a compromise on the wasm layer, we don't want to allow it to spoof consensus chains. Apps that only trusted consensus chains could now get a malicious message from wasm acting as one.